site/blog/tutorials/iasql-on-ec2.mdx

```diff
@@ -0,0 +1,284 @@
+- Set a security group for the instance to allow SSH, while disallowing unauthorized access to the dashboard
+
+<ThemedImage
+  alt="Overview of the resources to be deployed"
+  style={{width: '440'}}
+  sources={{
+    light: useBaseUrl('/img/iasql-on-ec2.png'),
+    dark: useBaseUrl('/img/iasql-on-ec2_dark.png'),
+  }}
+/>
+
+#### Install IaSQL Modules
+
+Let's start by installing the needed IaSQL modules. IaSQL has a modular design, and therefore to manage EC2 instances, we need to use [`aws_ec2` module](https://iasql.com/docs/modules/aws/aws_ec2/. Also, [`aws_ec2_metadata`](https://iasql.com/docs/modules/aws/aws_ec2_metadata/) allows us to get more info on our EC2 instance, like its public IP.
+```sql
+SELECT * FROM iasql_install('aws_ec2', 'aws_ec2_metadata');
+```
+
+#### Import SSH Key Pair Into AWS
+
+If you want to be able to access the instance through SSH, you can import your public key into AWS by executing the following IaSQL RPC:
+```sql
+SELECT * FROM key_pair_import('iasql', '<your-public-key-content>', 'us-east-1');
+```
+
+This will direct IaSQL to create an [AWS key pair](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/create-key-pairs.html#how-to-generate-your-own-key-and-import-it-to-aws) with the name `iasql` in the `us-east-1` region, and with your public key contents.
+
+:::note
+An IaSQL RPC is a function available in Postgres that upon calling, sends a request to the IaSQL engine to run a pre-defined logic and returns results to the Postgres console. In the above example [`key_pair_import`](https://iasql.com/docs/modules/aws/tables/aws_ec2_rpcs_import.KeyPairImportRpc/) is an IaSQL RPC that triggers the IaSQL engine to send a request to AWS for importing the public key.
+:::
+
+#### Create And Configure Security Group
+
+Now it's time to launch our EC2 instance. Before we go for the instance itself, let's first define the security group that's going to be attached to it. For this purpose, we'll need [`aws_security_group`](https://iasql.com/docs/modules/aws/aws_security_group/) module. But since `aws_ec2` module depends on it, it's already installed because IaSQL automatically installs the dependencies for modules upon an `iasql_install` call.
+
+<ThemedImage
+  alt="aws_ec2 module dependencies"
+  style={{width: '440'}}
+  sources={{
+    light: useBaseUrl('/img/module-dependency.png'),
+    dark: useBaseUrl('/img/module-dependency_dark.png'),
+  }}
+/>
+
+
+```sql
+INSERT INTO security_group (description, group_name)
+VALUES ('IaSQL security group', 'iasql_sg');
+
+WITH sg as (SELECT id
+            FROM security_group
+            WHERE group_name = 'iasql_sg')
+INSERT
+INTO security_group_rule (is_egress, ip_protocol, from_port, to_port, cidr_ipv4, description, security_group_id)
+VALUES (FALSE, 'tcp', 22, 22, '0.0.0.0/0', 'ssh port is accessible from all ips', (SELECT id FROM sg)),
+       (FALSE, 'tcp', 9879, 9879, '<your-ip>/0', 'open dashboard to just me', (SELECT id FROM sg)),
```

Comment:
Hmm, can't you just do
```suggestion
       (FALSE, 'tcp', 9879, 9879, '<your-ip>/0', 'open dashboard to just me', sg.id),
```

Reply:
That's what I initially tested. It would've been great if that worked, but it fails with this error:
```
[42P01] ERROR: missing FROM-clause entry for table "sg"
```

Reply:
indeed! Well maybe instead of `VALUES` you could use `SELECT` and then `FROM sg`, but well I think it is ok as is too!! just thinking alternatives

Reply:
How would you write it using `SELECT`? This is not working:
```sql
WITH sg as (SELECT id
            FROM security_group
            WHERE group_name = 'iasql_sg' LIMIT 1)
INSERT
INTO security_group_rule (is_egress, ip_protocol, from_port, to_port, cidr_ipv4, description, security_group_id)
SELECT (FALSE, 'tcp', 22, 22, '0.0.0.0/0', 'ssh port is accessible from all ips', sg.id),
       (FALSE, 'tcp', 9879, 9879, '0.0.0.0/0', 'open dashboard to just me', sg.id),
       (FALSE, 'tcp', 5432, 5432, '0.0.0.0/0', 'postgres port is accessible from all ips', sg.id),
       (TRUE, 'tcp', 1, 65535, '0.0.0.0/0', 'allow outgoing traffic to anywhere', sg.id)
FROM sg;
```
This won't work becuase it will treat the whole `(FALSE, 'tcp', 22, 22, '0.0.0.0/0', 'ssh port is accessible from all ips', sg.id)` as `is_egress`. Also if I remove the paranthesis then we can't insert multiple rows together.

Reply:
oh interesting, I was somehow expecting that query to work, but I guess it makes sense. Let's keep it as is!